### PR TITLE
Add shortcut with keyholder

### PR DESCRIPTION
**Pull request by AkaShark**
closed https://github.com/tisfeng/Easydict/issues/290

### CODE REVIEW

File: Easydict.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/ZipArchive/ZipArchive.git",
+      "state" : {
+        "revision" : "79d4dc9729096c6ad83dd3cee2b9f354d1b4ab7b",
+        "version" : "2.5.5"
+      }
+    }
+  ],
+  "version" : 2
+}
```
Comment (tisfeng) [inline Easydict.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:257]: I'm not familiar with Swift package configuration files, why are there two similar `Package.resolved` files here, do they serve the same purpose?

**Review by tisfeng — CHANGES_REQUESTED** [review]

File: Easydict.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 1 above)
Comment (AkaShark) [reply Easydict.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:257]: yeah, I noticed that just now, seems they are in different path one in `Easydict.xcwokspace` and other in `Easydict.xcodeproj`, but I never change them by manual, let me read some article  and try to explain them.

**Comment by tisfeng** [comment]
App crashed when I use the shortcut key `Opt+S` in old settings, I only set a new shortcut key `Opt+A` in new settings, `Opt+A` works fine.

```
Magnet/HotKeyCenter.swift:140: Assertion failed: Invalid hot key id
```

image

**Comment by AkaShark** [comment]
> App crashed when I use the shortcut key `Opt+S` in old settings, I only set a new shortcut key `Opt+A` in new settings, `Opt+A` works fine.
> 
> ```
> Magnet/HotKeyCenter.swift:140: Assertion failed: Invalid hot key id
> ```
> 
> image

okay, I repeated this crash, seems caused by the conflict between `keyholder` and `MAShortcut`, I also purposely used different Key values.... Let me see more details, I will fix this later.

**Comment by tisfeng** [comment]
The shortcut keys group looks inconsistent with the overall page layout. 

Maybe we can move the shortcut keys to a separate page and optimize the layout. For example, there is no need for dividing lines, and the shortcut key view needs to be left aligned.

image

**Comment by AkaShark** [comment]
> The shortcut keys group looks inconsistent with the overall page layout.
> 
> Maybe we can move the shortcut keys to a separate page and optimize the layout. For example, there is no need for dividing lines, and the shortcut key view needs to be left aligned.
> 
> image

That sounds like a good idea! I will move shortcut to another page, and optimize the layout

**Comment by tisfeng** [comment]
> MAShortcut



> App crashed when I use the shortcut key `Opt+S` in old settings, I only set a new shortcut key `Opt+A` in new settings, `Opt+A` works fine.
> 
> ```
> Magnet/HotKeyCenter.swift:140: Assertion failed: Invalid hot key id
> ```
> 
> image

I just fixed it in a simple way, and I didn't confirm the conflict between keyholder and MAShortcut.

```objc
    if (!EasydictNewAppManager.shared.enable) {
        [EZMenuItemManager.shared setup];
        [EZShortcut setup];
    } else {
        [Shortcut setupShortcut];
    }
```

**Comment by AkaShark** [comment]
> > MAShortcut
> 
> > App crashed when I use the shortcut key `Opt+S` in old settings, I only set a new shortcut key `Opt+A` in new settings, `Opt+A` works fine.
> > ```
> > Magnet/HotKeyCenter.swift:140: Assertion failed: Invalid hot key id
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > image
> 
> I just fixed it in a simple way, and I didn't confirm the conflict between keyholder and MAShortcut.
> 
> ```objc
>     if (!EasydictNewAppManager.shared.enable) {
>         [EZMenuItemManager.shared setup];
>         [EZShortcut setup];
>     } else {
>         [Shortcut setupShortcut];
>     }
> ```

You are right, that is a smart way !

**Comment by tisfeng** [comment]
This shortcut key layout looks weird.

image

**Comment by AkaShark** [comment]
> This shortcut key layout looks weird.
> 
> image

yeah, that is my [problem](https://github.com/tisfeng/Easydict/pull/349#issue-2092403336), in UIKit we can use `NSMenuItem` to add `keyEquicalent`  and `keyEquivalentModifierMask` 
CleanShot 2024-01-21 at 20 59 57@2x
But in SwiftUI I can't find the property 
CleanShot 2024-01-21 at 21 01 23@2x
If I use `.keyboardShortcut` user can't add a double modifier key. I tried to use HStack but it did seem not good 😭

**Comment by tisfeng** [comment]
I tried this code, looks good.

image

File: Easydict/NewApp/Feature/Shortcut/Shortcut.swift [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+//    }
+//
+//    func fetchShortcutKeyEventModifiers(_ type: ShortcutType) -> EventModifiers {
+//        guard let keyCombo = fetchShortcutKeyCombo(type) else { return [] }
+//        return [.command, .command]
+//    }
+
+    func fetchShortcutKeyStr(_: ShortcutType) -> String {
+        guard let keyCombo = restoreInputTranslate() else { return "" }
+        return shortcutKeyStr(keyCombo)
```
Comment (tisfeng) [inline Easydict/NewApp/Feature/Shortcut/Shortcut.swift:174]: We don't need a shortcut key string, key combo is enough.

**Review by tisfeng — CHANGES_REQUESTED** [review]

**Comment by AkaShark** [comment]
> I tried this code, looks good.
> 
> image

Haha! you are right! I forget to try use ’⌥‘, very nice way! wait me for moments

**Comment by CanglongCl** [comment]
I tried add the keyboard shortcut. It worked well. 

However, it seems like deleting shortcut is not working. After I click the X button, it shows the shortcut is canceled. But the deleted shortcut still trigger opening the window.

**Comment by CanglongCl** [comment]
Btw, I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.

**Comment by CanglongCl** [comment]
The menu bar doesn't update after the setting of shortcut. Only after we restart the app, it updates.

**Review by CanglongCl — CHANGES_REQUESTED** [review]
I tried add the keyboard shortcut. It worked well.

Here is some issues:

It seems like deleting shortcut is not working. After I click the X button, it shows the shortcut is canceled. But the deleted shortcut still trigger opening the window.
I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.
The menu bar doesn't update after the setting of shortcut. Only after we restart the app, it updates.
For UI, this packages' UI is not kinda clearly. We should add more instruction to show the shortcut can be edited just like that in old app.

**Comment by CanglongCl** [comment]
> I tried add the keyboard shortcut. It worked well.
> 
> Here is some issues:
> 
> 1. It seems like deleting shortcut is not working. After I click the X button, it shows the shortcut is canceled. But the deleted shortcut still trigger opening the window.
> 2. I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.
> 3. The menu bar doesn't update after the setting of shortcut. Only after we restart the app, it updates.
> 4. For UI, this packages' UI is not kinda clearly. We should add more instruction to show the shortcut can be edited just like that in old app.

I've resolved the issue 1 & 3 and did some refactor to utilize `Defaults` with `KeyCombo` with a commit. Please review it. @AkaShark

File: Easydict/NewApp/Feature/Shortcut/Shortcut.swift [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            guard let keyCombo = restoreShowMiniWindow() else { return nil }
+            return keyCombo
+        }
+    }
+}
+
+/// can't using keyEquivalent and EventModifiers in SwiftUI MenuItemView direct, because item
+/// keyboardShortcut not support double modifier key but can use ⌥ as character
+extension View {
+    public func keyboardShortcut(_ type: ShortcutType) -> some View {
```
Comment (CanglongCl) [inline Easydict/NewApp/Feature/Shortcut/Shortcut.swift]: `AnyView` is not recommended. Try to replaced with using `@ViewBuilder` modifier. 

```swift
@ViewBuilder
func keyboardShortcut(_ type: ShortcutType) -> some View {
        if let keyCombo = ... {
                self.keyboardShortcut(...)
        } else {
                self
        }
}
```

**Comment by AkaShark** [comment]
> > I tried add the keyboard shortcut. It worked well.
> > Here is some issues:
> > 
> > 1. It seems like deleting shortcut is not working. After I click the X button, it shows the shortcut is canceled. But the deleted shortcut still trigger opening the window.
> > 2. I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.
> > 3. The menu bar doesn't update after the setting of shortcut. Only after we restart the app, it updates.
> > 4. For UI, this packages' UI is not kinda clearly. We should add more instruction to show the shortcut can be edited just like that in old app.
> 
> I've resolved the issue 1 & 3 and did some refactor to utilize `Defaults` with `KeyCombo` with a commit. Please review it. @AkaShark

Thanks for your refactor, `Defaults` is powerful! I got it.

**Comment by AkaShark** [comment]
> > I tried add the keyboard shortcut. It worked well.
> > Here is some issues:
> > 
> > 1. It seems like deleting shortcut is not working. After I click the X button, it shows the shortcut is canceled. But the deleted shortcut still trigger opening the window.
> > 2. I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.
> > 3. The menu bar doesn't update after the setting of shortcut. Only after we restart the app, it updates.
> > 4. For UI, this packages' UI is not kinda clearly. We should add more instruction to show the shortcut can be edited just like that in old app.
> 
> I've resolved the issue 1 & 3 and did some refactor to utilize `Defaults` with `KeyCombo` with a commit. Please review it. @AkaShark

@CanglongCl  Can I ask you some questions, I am not familiar with SwiftUI. Why issues 3 are well done by your commit. Is contribution by `@ViewBuilder`? I am not find change in `GeneralKeyHolderWrapper.swift`

**Comment by CanglongCl** [comment]
> > > I tried add the keyboard shortcut. It worked well.
> > > Here is some issues:
> > > 
> > > 1. It seems like deleting shortcut is not working. After I click the X button, it shows the shortcut is canceled. But the deleted shortcut still trigger opening the window.
> > > 2. I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.
> > > 3. The menu bar doesn't update after the setting of shortcut. Only after we restart the app, it updates.
> > > 4. For UI, this packages' UI is not kinda clearly. We should add more instruction to show the shortcut can be edited just like that in old app.
> > 
> > 
> > I've resolved the issue 1 & 3 and did some refactor to utilize `Defaults` with `KeyCombo` with a commit. Please review it. @AkaShark
> 
> @CanglongCl Can I ask you some questions, I am not familiar with SwiftUI. Why issues 3 are well done by your commit. Is contribution by `@ViewBuilder`? I am not find change in `GeneralKeyHolderWrapper.swift`

I did some further refactor with a `ViewModifier` where I can add a view like struct in modifier. In fact, while you are writing `var body: some View`, the `body` in `View` protocol is declared with `@ViewBuilder` so you are actually using it without explicitly see it (same for `func content() -> some View` in `ViewModifier` in my commit). 

For `ViewBuilder`, it is a sugar in Swift called `resultBuilder`. You can try read related documents to find more information. Here is a good blog explaining it. 

[ViewBuilder 研究（上）—— 掌握 Result builders](https://fatbobman.com/zh/posts/viewbuilder1/)

**Comment by tisfeng** [comment]
> 2. I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.

It seems that when using the new `EasydictApp.swift` to launch, the shortcuts in the previous `Main.storyboard` are not loaded, and shortcuts like `Cmd+P` for pinning the window do not work. In addition, we can also move these in-app shortcuts to the new shortcut tab and allow users to modify them.

I'll open another issue for this problem, which can be resolved later.

image

**Comment by AkaShark** [comment]
> > 2. I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.
> 
> It seems that when using the new `EasydictApp.swift` to launch, the shortcuts in the previous `Main.storyboard` are not loaded, and shortcuts like `Cmd+P` for pinning the window do not work. In addition, we can also move these in-app shortcuts to the new shortcut tab and allow users to modify them.
> 
> I opened a issue #352 for this problem, which can be resolved later.
> 
> image

When I add the shortcut form `Main.storyboard` to `ShortcutTab.swift` I find the `Main(String)` has the key, but the key seems strange Can I use this key directly?
CleanShot 2024-01-24 at 14 22 30@2x

**Comment by tisfeng** [comment]
Can this PR be finished this weekend?

**Comment by AkaShark** [comment]
> Can this PR be finished this weekend?

sure, I will optimize this PR in this weekend

**Comment by AkaShark** [comment]
> I'm planning to release a new version this weekend, will this PR make it in time?

yeah, I will PR in tomorrow. Sorry, there are some things I can't get out of in last weekend

**Comment by tisfeng** [comment]
OK, it's okay.

**Comment by AkaShark** [comment]
> The shortcut keys group looks inconsistent with the overall page layout.
> 
> Maybe we can move the shortcut keys to a separate page and optimize the layout. For example, there is no need for dividing lines, and the shortcut key view needs to be left aligned.
> 
> image

Already move shortcut into new page and optimize UI like this
CleanShot 2024-01-30 at 14 46 21@2x

**Comment by CanglongCl** [comment]
> > The shortcut keys group looks inconsistent with the overall page layout.
> > Maybe we can move the shortcut keys to a separate page and optimize the layout. For example, there is no need for dividing lines, and the shortcut key view needs to be left aligned.
> > image
> 
> Already move shortcut into new page and optimize UI like this CleanShot 2024-01-30 at 14 46 21@2x

I will prefer to align the text (with out `:`) left and shortcuts right, which is consistent with other tabs.

**Comment by AkaShark** [comment]
> I tried add the keyboard shortcut. It worked well.
> 
> Here is some issues:
> 
> 1. It seems like deleting shortcut is not working. After I click the X button, it shows the shortcut is canceled. But the deleted shortcut still trigger opening the window.
> 2. I think we should add some limitation on shortcut options. For example, cmd+q and cmd+, should not be set as shortcut.
> 3. The menu bar doesn't update after the setting of shortcut. Only after we restart the app, it updates.
> 4. For UI, this packages' UI is not kinda clearly. We should add more instruction to show the shortcut can be edited just like that in old app.

For 2 I add system and menu item shortcut conflict validator, show alter when happen shortcut conflict like this 
CleanShot 2024-01-30 at 14 50 06@2x
For 4 I have no idea, but I think instructions like this will be Ok, I will commit later.
CleanShot 2024-01-30 at 14 50 55@2x

**Comment by Jerry23011** [comment]
It seems that `Shortcut` is marked as Stale in String Catalog, the string seems to be `shortcut`?
截屏2024-01-29 23 45 13

**Comment by AkaShark** [comment]
> It seems that `Shortcut` is marked as Stale in String Catalog, the string seems to be `shortcut`? 截屏2024-01-29 23 45 13

ok, I have removed unused strings in `.xcstrings`.

**Comment by Jerry23011** [comment]
I think we also need translation for simplified Chinese `shortcut`

**Comment by Jerry23011** [comment]
It’s there now, thanks

**Review by Jerry23011 — DISMISSED** [review]

**Comment by AkaShark** [comment]
This feature is ready. Please  review the change.

File: Easydict/NewApp/Feature/Shortcut/Shortcut.swift [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    func reset() {
+        // Reset all properties to default value
+    }
+}
+
+// restore shortcut
+extension Shortcut {
+    func restoreShortcut() {
+        if let inputTranslateKeyCombo = Defaults[.inputShortcut] {
+            bindingShortCut(keyCombo: inputTranslateKeyCombo, type: .inputTranslate)
```
[the remark this answers is not in the document]
Comment (phlpsong) [reply Easydict/NewApp/Feature/Shortcut/Shortcut.swift:61]: And also it would be better to rename it to `bindingShortcut`.

**Comment by phlpsong** [comment]
There are some localized strings marked as `need_review`, do we need to update this? 

Screenshot 2024-01-31 at 22 17 12

**Comment by AkaShark** [comment]
> There are some localized strings marked as `need_review`, do we need to update this?

yeah, I think we need to mark the changes that we confirm as reviewed. I have marked this as reviewed.

**Review by phlpsong — CHANGES_REQUESTED** [review]

**Review by phlpsong — DISMISSED** [review]

**Review by tisfeng — APPROVED** [review]
Looks good.

File: Easydict.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 1 above)
Comment (tisfeng) [reply Easydict.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:257]: How about this?

**Review by phlpsong — APPROVED** [review]

**Review by tisfeng — APPROVED** [review]

**Comment by tisfeng** [comment]
I will merge this PR soon if there are no other questions.

**Review by CanglongCl — APPROVED** [review]

**Pull request merged**